Move dotnet.cmd/sh helper scripts into the repo root

ViktorHofer opened the pull request:
Several people expressed their interest in having the dotnet.cmd/ps1/sh helper scripts, which invoke the correct dotnet, in the repo root. This also improves the local developer workflow as it offers an alternative to invoking the globally installed dotnet.

ViktorHofer committed:
Move dotnet helper scripts to repo root

ViktorHofer committed:
Delete broken script which invokes dotnet.cmd

ViktorHofer committed:
Fix -x permissions and optionally install cli

jkotas commented on dotnet.cmd:
```diff
@@ -1,38 +1,20 @@
-echo running: %PS_COMMAND%
 for /f "delims=" %%l in ('%PS_COMMAND%') do set "__dotnetDir=%%l"
 
-if NOT [%ERRORLEVEL%] == [0] (
```
Is it a good idea to delete this error handling? If the SDK fails to install e.g. due the network error, is average person going to be able to figure that there was a network error?

ViktorHofer replied:
The arcade scripts handle such errors and print in failure cases.

jkotas replied:
> The arcade scripts handle such errors and print in failure cases.

I do not see any errors printed by arcade scripts. Here is what I see with disconnected network:

```
C:\runtime>git clean -x -d -f .
Removing .dotnet/
Removing artifacts/

C:\runtime>dotnet
'" \dotnet.exe"' is not recognized as an internal or external command,
operable program or batch file.
```

ViktorHofer replied:
Yes, that's because @echo off turns off all logging which is incorrect. Thanks for pointing that out.

ViktorHofer replied:
Submitted a fix.

ViktorHofer committed:
Don't swallow InitializeDotNetCli output

Introducing a dotnet.ps1 wrapper script which invokes the bootstrapping
logic so that output in dotnet.cmd isn't swallowed.

ViktorHofer committed:
Merge branch 'DotNetHelperScripts' of https://github.com/ViktorHofer/runtime into DotNetHelperScripts

wjk commented on dotnet.sh:
```diff
@@ -0,0 +1,27 @@
+__dotnetDir=${_InitializeDotNetCli}
+
+dotnetPath=${__dotnetDir}/dotnet
+${dotnetPath} "$@"
```
```suggestion
${dotnetPath} "$@"
exec ${dotnetPath} "$@"
```

wjk reviewed (commented):
Just one thought, mostly stylistic in nature.

trylek reviewed (approved):
Looks good, thanks for cleaning this up!

jkotas commented on dotnet.cmd:
```diff
@@ -0,0 +1,2 @@
+@echo off
+powershell -ExecutionPolicy ByPass -NoProfile -command "& """%~dp0dotnet.ps1""" %*"
```
I do not think we should be executing dotnet indirectly via powershell. These wrappers over wrappers just cause pain.

jkoritzinsky replied:
The Arcade powershell script also locates the dotnet cli as well as downloading it if it isn't present. The previous version of dotnet.cmd also called into powershell to do the same thing, but in a less straight-forward manner.

jkotas replied:
I am fine with calling powershell to download and locale dotnet.exe. I would like to see the actual invocation of dotnet.exe to be done directly from the .cmd/.sh files.

jkotas replied:
I believe that the diagnosability is indirectly proportion to the max depth of the process tree. (My memories of `run` tool that added an extra process layer everywhere and made problems so much harder to troubleshoot are still too fresh.)

We are not introducing the extra process layer on Unix and invoking dotnet directly from the .sh script. Can we do it the same way on Windows too?

ViktorHofer replied:
Thanks for letting us know, I will explore other options without the indirection. On Unix this is fine as the arcade scripts can be `dot` sourced into the caller as the environment and the shell engine is the same. On Windows this is more tricky as their is no standard mechanism to share variables between cmd and powershell scripts besides stdin/stdout, registry, global vars, etc. I will probably find another way to do this.

ViktorHofer replied:
> I will probably find another way to do this.

After researching a bit I'm not less sure about this.

jkotas replied:
What is wrong with the `for /f ...` used in the current script (https://github.com/dotnet/runtime/blob/master/src/coreclr/dotnet.cmd#L21)? That's one of the best way to do this, as far as I know.

ViktorHofer replied:
> What is wrong with the for /f ... used in the current script

That all error handling and stdout in the arcade script is swallowed. The arcade function is not responsible for returning the path to the SDK folder but also to add telemetry markers when being used in CI. There must be a better way to do this.

ViktorHofer replied:
@jkotas I addressed your concerns, the calling script now invokes the host. The powershell script now has the following flow:

1. Retrieve the SDK folder
2a. Call the host if the script was invoked in a normal way
2b. Call the dotnet.cmd script if the `Get-DotNetPath` function was invoked.

Similar for the dotnet.cmd script which detects if the SDK directory is passed in as argument or not.

ViktorHofer replied:
Oh I believe the issue with process tree additional depth is still present.

ViktorHofer replied:
OK one more iteration, I'm now using a temporary file to store the location similar to what @trylek tried initially. This seems to be the cleanest option.

jkotas replied:
> That all error handling and stdout in the arcade script is swallowed

`for /f ... ` won't swallow error handling. If the command prints something to stderr, it gets passed through.

Are you saying that Arcade scripts print errors to stdout instead of stderr?

ViktorHofer replied:
As mentioned the Arcade scripts do pipeline logging as well, ie https://github.com/dotnet/runtime/blob/285911db951b2721c46f4e5d65bf1b419ae07c03/eng/common/tools.ps1#L157-L160.

> for /f ... won't swallow error handling. If the command prints something to stderr, it gets passed through.

It does, it swallows the default powershell exceptions, I just tried that locally when checking out my first commit in this branch. If I keep the `@echo off` I don't get any output. If I remove the `@echo off` I get the following output:

```
C:\git\runtime2>.\dotnet.cmd

C:\git\runtime2>setlocal

C:\git\runtime2>set Platform=

C:\git\runtime2>set DOTNET_MULTILEVEL_LOOKUP=0

C:\git\runtime2>set DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1

C:\git\runtime2>set PS_DOTNET_INSTALL_SCRIPT=". C:\git\runtime2\eng\common\tools.ps1; InitializeDotNetCli($true)"

C:\git\runtime2>set "PS_COMMAND=powershell -NoProfile -ExecutionPolicy unrestricted -Command ". C:\git\runtime2\eng\common\tools.ps1; InitializeDotNetCli($true)""

C:\git\runtime2>for /F "delims=" %l in ('powershell -NoProfile -ExecutionPolicy unrestricted -Command ". C:\git\runtime2\eng\common\tools.ps1; InitializeDotNetCli($true)"') do set "__dotnetDir=%l"

C:\git\runtime2>set "__dotnetDir=Invoke-WebRequest : The remote name could not be resolved: 'dot.net'"

C:\git\runtime2>set "__dotnetDir=At C:\git\runtime2\eng\common\tools.ps1:170 char:5"

C:\git\runtime2>set "__dotnetDir=+     Invoke-WebRequest "https://dot.net/$dotnetInstallScriptVersion/do ..."

C:\git\runtime2>set "__dotnetDir=+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"

C:\git\runtime2>set "__dotnetDir=    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc "

C:\git\runtime2>set "__dotnetDir=   eption"

C:\git\runtime2>set "__dotnetDir=    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand"

C:\git\runtime2>set "__dotnetDir= "

C:\git\runtime2>set "dotnetPath= \dotnet.exe"

C:\git\runtime2>call " \dotnet.exe"
'" \dotnet.exe"' is not recognized as an internal or external command,
operable program or batch file.

C:\git\runtime2>.\dotnet.cmd
```

jkotas replied:
> It does, it swallows the default powershell exceptions

That's because of powershell seems to like to print exceptions by default to stdout, not to stderr.

If I create powershell script `oops.ps1` that prints an error to stderr:
```
[Console]::Error.WriteLine('oops')
```

And call it from a `test.bat` like this:

```
@echo  off
for /f "delims=" %%l in ('powershell -NoProfile -ExecutionPolicy unrestricted -Command .\oops.ps1') do set "__dotnetDir=%%l"
```

I do see the error when invoking the .bat file:

```
D:\repro>test.bat
oops
```

ViktorHofer committed:
Invoke dotnet SDK in the calling script

Add the powershell script back and invoke the SDK in the corresponding
root script.

ViktorHofer committed:
Use temporary file to store sdk location

To avoid a nested process chain in which the inner process then invokes
the host, we instead create a temporary file to store the sdk location.

ViktorHofer committed:
Guard against multiple invocations

ViktorHofer committed:
Guard against invocation error

jkotas commented on dotnet.ps1:
```diff
@@ -0,0 +1,28 @@
+function Get-DotNetPath() {
+  $sdkCacheFile = (Join-Path $ToolsetDir 'sdk.txt')
+  Try {
```
Would it be better to instead check whether `sdk.txt` exists to avoid the try/catch?

ViktorHofer replied:
the Try Catch here guards against concurrent write to the file. In case multiple processes want to write to that file at the same time. I could introduce a Mutex but I think that would be an overkill. I should add a comment.

jkotas replied:
This does not fix the race condition reliably. If you are worried about race conditions like this, the best way to handle them is to create a temp file and then atomically rename the file.

ViktorHofer replied:
But doesn't the calling script `dotnet.cmd` need to know the location of that temp script as well? Maybe I'm misunderstanding something.

jkotas replied:
I meant something like this:

```
$sdkCacheFileTemp = CommonLibrary\Get-TempPathFilename -Path  $ToolsetDir
Set-Content -Path $sdkCacheFileTemp -Value $dotnetPath

Try {
    Rename-Item -Path $sdkCacheFileTemp 'sdk.txt'
} Catch {
    # Somebody beat us
}
```

ViktorHofer replied:
Got it, thanks.

ViktorHofer commented:
I could submit a PR to arcade which adds an option to create a file with the path to the SDK and get rid of the dotnet.ps1 here.

ViktorHofer committed:
Add temp file logic to tools.ps1

ViktorHofer commented:
Merging as the [arcade PR](https://github.com/dotnet/arcade/pull/4531) is merged which means we won't loose the changes that I made to the tools.ps1 file.
